service/storage_service.cc

```diff
@@ -316,6 +316,22 @@ future<> storage_service::topology_state_load() {
         co_return *ip;
     };
 
+    if (_topology_state_machine._topology.has_complete_feature_information()) {
+        const auto enabled_features = _topology_state_machine._topology.calculate_enabled_features();
+        // We must check here that the node supports all features enabled
+        // in the cluster so far. If we won't do that then there is a risk
+        // that further commands might depend on some features that the current
+        // node doesn't understand and it might risk misinterpreting them.
+        //
+        // An interesting and useful side effect of those checks is that
+        // performing a successful read barrier makes sure that the local node
+        // supports all cluster's features.
+        const auto supported_by_me = _feature_service.supported_feature_set();
+        service::topology::check_if_enabled_features_are_known(supported_by_me, enabled_features);
+        std::set<std::string_view> enabled_features_v = boost::copy_range<std::set<std::string_view>>(enabled_features);
+        co_await _feature_service.enable(std::move(enabled_features_v));
```

Comment:
We should optimize this somehow so we don't reenable all features on every topology commands, which involves disk writes and a bunch of log messages which are just going to spam the logs unnecessarily. We should enable only whatever new features have appeared. (The optimization looks like it belongs inside `feature_service` more than here though)

Reply:
It's not that bad, if a feature is already enabled then `_feature_service.enable()` won't enable them again (i.e. trigger callbacks, produce log messages or do disk writes).

The only costly thing that it does unconditionally is spawning a `seastar::async`. I guess this could be avoided if no new features are to be enabled, but I'm not sure whether it's worth it.